libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -522,91 +522,102 @@ bool ContractCompiler::visit(InlineAssembly const& _inlineAssembly)
-							);
-						for (unsigned i = 0; i < variable->type()->sizeOnStack(); ++i)
-							_assembly.append(dupInstruction(stackDiff));
+						solAssert(location.second == 0, "Intra-slot offest assumed to be zero.");
+						_assembly.append(location.first);
 					}
 					else
 					{
-						solAssert(m_context.isStateVariable(variable), "Invalid variable type.");
-						auto const& location = m_context.storageLocationOfVariable(*variable);
-						if (!variable->type()->isValueType())
-						{
-							solAssert(location.second == 0, "Intra-slot offest assumed to be zero.");
-							_assembly.append(location.first);
-						}
-						else
-						{
-							_assembly.append(location.first);
-							_assembly.append(u256(location.second));
-						}
+						_assembly.append(location.first);
+						_assembly.append(u256(location.second));
 					}
 				}
-				else if (auto contract = dynamic_cast<ContractDefinition const*>(decl))
-				{
-					solAssert(contract->isLibrary(), "");
-					_assembly.appendLibraryAddress(contract->fullyQualifiedName());
-				}
-				else
-					solAssert(false, "Invalid declaration type.");
-			} else {
-				// lvalue context
-				auto variable = dynamic_cast<VariableDeclaration const*>(decl);
-				solAssert(
-					!!variable && m_context.isLocalVariable(variable),
-					"Can only assign to stack variables in inline assembly."
+			}
+			else if (auto contract = dynamic_cast<ContractDefinition const*>(decl))
+			{
+				solAssert(contract->isLibrary(), "");
+				_assembly.appendLibraryAddress(contract->fullyQualifiedName());
+			}
+			else
+				solAssert(false, "Invalid declaration type.");
+		}
+		else
+		{
+			// lvalue context
+			auto variable = dynamic_cast<VariableDeclaration const*>(decl);
+			solAssert(
+				!!variable && m_context.isLocalVariable(variable),
+				"Can only assign to stack variables in inline assembly."
+			);
+			unsigned size = variable->type()->sizeOnStack();
+			int stackDiff = _assembly.deposit() - m_context.baseStackOffsetOfVariable(*variable) - size;
+			if (stackDiff > 16 || stackDiff < 1)
+				BOOST_THROW_EXCEPTION(
+					CompilerError() <<
+					errinfo_sourceLocation(_inlineAssembly.location()) <<
+					errinfo_comment("Stack too deep, try removing local variables.")
 				);
-				unsigned size = variable->type()->sizeOnStack();
-				int stackDiff = _assembly.deposit() - m_context.baseStackOffsetOfVariable(*variable) - size;
-				if (stackDiff > 16 || stackDiff < 1)
-					BOOST_THROW_EXCEPTION(
-						CompilerError() <<
-						errinfo_sourceLocation(_inlineAssembly.location()) <<
-						errinfo_comment("Stack too deep, try removing local variables.")
-					);
-				for (unsigned i = 0; i < size; ++i) {
-					_assembly.append(swapInstruction(stackDiff));
-					_assembly.append(Instruction::POP);
-				}
+			for (unsigned i = 0; i < size; ++i)
+			{
+				_assembly.append(swapInstruction(stackDiff));
+				_assembly.append(Instruction::POP);
```

Comment:
This code is generated for `x := <expression>` or ` =: x`. You are asking what happens if, at the point where code is generated for the actual assignment (not the expression that might be involved), the value that is assigned reaches below the stack height of some variable.

In the first case, it is only possible to end up in such a situation if we have been in such a situation already before the assignment. In the second case, you could consume stack elements and reach below the highest named stack slot, but the second case is not part of the "strict mode" we were talking about before where you can only use functional statements that do not leave items on the stack.

If some variable would point at that position and you would use that variable before the position is filled again, an error similar to the one in (green) line 606 would be generated.

I'm fine with adding more checks (that also check whether you are using "strict mode") in a future PR that would also check that the stack height never reaches below the highest named stack slot.